File: solutions/typescript/0080.ts
/**
 * 80. 删除有序数组中的重复项 II
 * 给定有序数组 nums，原地删除重复元素使每个元素最多出现两次，
 * 返回新长度。前 length 项即为结果。
 * 快慢双指针：slow 表示下一个写入位置，
 * 通用条件：slow < 2 或 nums[fast] !== nums[slow - 2]。
 */
function removeDuplicates(nums: number[]): number {
    const k = 2;
    let slow = 0;
    for (let fast = 0; fast < nums.length; fast++) {
        if (slow < k || nums[fast] !== nums[slow - k]) {
            nums[slow] = nums[fast];
            slow++;
        }
    }
    return slow;
}

export default removeDuplicates;

File: solutions/rust/0080.rs
/// 80. 删除有序数组中的重复项 II
/// 快慢双指针：每个元素至多出现两次。
/// 通用判断：slow < k 或 nums[fast] != nums[slow-k]（k=2）。
pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
    let k: usize = 2;
    let mut slow: usize = 0;
    for fast in 0..nums.len() {
        if slow < k || nums[fast] != nums[slow - k] {
            nums[slow] = nums[fast];
            slow += 1;
        }
    }
    slow as i32
}

impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        remove_duplicates(nums)
    }
}